File: functional-tests/functional/android/src/test/java/com/here/android/test/MultipleInheritanceTest.java
/*
 * Copyright (C) 2016-2021 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.android.test;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;

import android.os.Build;
import com.here.android.RobolectricApplication;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = Build.VERSION_CODES.M, application = RobolectricApplication.class)
public class MultipleInheritanceTest {

  private static class MultiInterfaceImpl implements MultiInterface {
    @Override
    public void childFunction() { }

    @Override
    public String getChildProperty() {
      return "";
    }

    @Override
    public void setChildProperty(final String value) { }

    @Override
    public void parentFunction() { }

    @Override
    public String getParentProperty() {
      return "";
    }

    @Override
    public void setParentProperty(final String value) { }

    @Override
    public String parentFunctionLight() {
      return "java face faces";
    }

    @Override
    public String getParentPropertyLight() {
      return "";
    }

    @Override
    public void setParentPropertyLight(final String value) { }
  }

  @Test
  public void fromCppSendUpcastSucceeds() {
    MultiClass instance = MultipleInheritanceFactory.getMultiClass();

    boolean result = instance instanceof NarrowInterface;

    assertTrue(result);
  }

  @Test
  public void fromCppSendDowncastFails() {
    NarrowInterface instance = MultipleInheritanceFactory.getMultiClassAsNarrow();

    boolean result = instance instanceof MultiClass;

    assertFalse(result);
  }

  @Test
  public void fromCppSendTwiceEquals() {
    NarrowInterface instance1 = MultipleInheritanceFactory.getMultiClassSingleton();
    NarrowInterface instance2 = MultipleInheritanceFactory.getMultiClassSingleton();

    boolean result = instance1 == instance2;

    assertTrue(result);
  }

  @Test
  public void fromCppRoundTripNotEquals() {
    NarrowInterface instance = MultipleInheritanceFactory.getMultiClassSingleton();

    boolean result = MultipleInheritanceChecker.checkSingletonEquality(instance);

    assertFalse(result);
  }

  @Test
  public void fromCppRoundTripWithUpcastNotEquals() {
    MultiClass uncastInstance = MultipleInheritanceFactory.getMultiClass();
    NarrowInterface instance = (NarrowInterface)uncastInstance;

    boolean result = MultipleInheritanceChecker.checkSingletonEquality(instance);

    assertFalse(result);
  }

  @Test
  public void fromJavaSendUpcastSucceeds() {
    MultiInterface instance = new MultiInterfaceImpl();

    boolean result = MultipleInheritanceChecker.checkIsNarrow(instance);

    assertTrue(result);
  }

  @Test
  public void fromJavaSendDowncastFails() {
    MultiInterface uncastInstance = new MultiInterfaceImpl();
    NarrowInterface instance = (NarrowInterface)uncastInstance;

    boolean result = MultipleInheritanceChecker.checkIsMultiInterface(instance);

    assertFalse(result);
  }

  @Test
  public void fromJavaSendTwiceEquals() {
    NarrowInterface instance = new MultiInterfaceImpl();

    boolean result = MultipleInheritanceChecker.checkNarrowEquality(instance, instance);

    assertTrue(result);
  }

  @Test
  public void fromJavaRoundTripEquals() {
    MultiInterface uncastInstance = new MultiInterfaceImpl();
    NarrowInterface instance = (NarrowInterface)uncastInstance;

    boolean result = uncastInstance == MultipleInheritanceChecker.narrowRoundTrip(instance);

    assertTrue(result);
  }

  @Test
  public void fromJavaRoundTripWithUpcastNotEquals() {
    MultiInterface instance = new MultiInterfaceImpl();

    boolean result =
        instance == MultipleInheritanceFactory.upcastMultiInterfaceToNarrow(instance);

    assertFalse(result);
  }
}

File: functional-tests/functional/android-kotlin/src/test/kotlin/com/here/android/test/MultipleInheritanceTest.kt
/*
 * Copyright (C) 2016-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.android.test

import com.here.android.RobolectricApplication
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(application = RobolectricApplication::class)
class MultipleInheritanceTest {

    class MultiInterfaceImpl : MultiInterface {
        override fun childFunction() {}

        override var childProperty: String
            get() = ""
            set(_) {}

        override fun parentFunction() {}

        override var parentProperty: String
            get() = ""
            set(_) {}

        override fun parentFunctionLight(): String {
            return "kotlin face faces"
        }

        override var parentPropertyLight: String
            get() = ""
            set(_) {}
    }

    @org.junit.Test
    fun fromCppSendUpcastSucceeds() {
        val instance: MultiClass? = MultipleInheritanceFactory.getMultiClass()
        assertTrue(instance is NarrowInterface)
    }

    @org.junit.Test
    fun fromCppSendDowncastFails() {
        val instance: NarrowInterface = MultipleInheritanceFactory.getMultiClassAsNarrow()
        assertFalse(instance is MultiClass)
    }

    @org.junit.Test
    fun fromCppSendTwiceEquals() {
        val instance1: NarrowInterface = MultipleInheritanceFactory.getMultiClassSingleton()
        val instance2: NarrowInterface = MultipleInheritanceFactory.getMultiClassSingleton()

        assertTrue(instance1 === instance2)
    }

    @org.junit.Test
    fun fromCppRoundTripNotEquals() {
        val instance: NarrowInterface = MultipleInheritanceFactory.getMultiClassSingleton()
        assertFalse(MultipleInheritanceChecker.checkSingletonEquality(instance))
    }

    @org.junit.Test
    fun fromCppRoundTripWithUpcastNotEquals() {
        val uncastInstance: MultiClass = MultipleInheritanceFactory.getMultiClass()
        val instance: NarrowInterface = uncastInstance

        assertFalse(MultipleInheritanceChecker.checkSingletonEquality(instance))
    }

    @org.junit.Test
    fun fromKotlinSendUpcastSucceeds() {
        val instance: MultiInterface = MultiInterfaceImpl()
        val result: Boolean = MultipleInheritanceChecker.checkIsNarrow(instance)

        assertTrue(result)
    }

    @org.junit.Test
    fun fromKotlinSendDowncastFails() {
        val uncastInstance: MultiInterface = MultiInterfaceImpl()
        val instance: NarrowInterface = uncastInstance

        val result: Boolean = MultipleInheritanceChecker.checkIsMultiInterface(instance)
        assertFalse(result)
    }

    @org.junit.Test
    fun fromKotlinSendTwiceEquals() {
        val instance: NarrowInterface = MultiInterfaceImpl()
        val result: Boolean = MultipleInheritanceChecker.checkNarrowEquality(instance, instance)

        assertTrue(result)
    }

    @org.junit.Test
    fun fromKotlinRoundTripEquals() {
        val uncastInstance: MultiInterface = MultiInterfaceImpl()
        val instance: NarrowInterface = uncastInstance

        assertTrue(uncastInstance === MultipleInheritanceChecker.narrowRoundTrip(instance))
    }

    @org.junit.Test
    fun fromKotlinRoundTripWithUpcastNotEquals() {
        val instance: MultiInterface = MultiInterfaceImpl()

        assertFalse(instance === MultipleInheritanceFactory.upcastMultiInterfaceToNarrow(instance))
    }
}